### PR TITLE
Add batch async shard fetch transport action for replica #8218

### PR DESCRIPTION
**Pull request by sudarshan-baliga**
### Description
This pull request is part of the improvement https://github.com/opensearch-project/OpenSearch/issues/5098
This PR for adding the transport action for the async fetching of the shard information for batch of replica shards from the nodes in a single request per node. Code is inspired by the existing transport action [TransportNodesListShardStoreMetadata.java](https://github.com/opensearch-project/OpenSearch/compare/main...Gaurav614:OpenSearch:async-shard-fetch-rsatransport?expand=1#diff-79590563683e78d1a25985fdaee8adf7efabdc8792f6ec028741fb1d1d9ca5ad) which fetches just one shard information from the node per request.
* Added Transport action [TransportNodesBatchListShardStoreMetadata.java](https://github.com/opensearch-project/OpenSearch/compare/main...Gaurav614:OpenSearch:async-shard-fetch-rsatransport?expand=1#diff-04c725a5fe9a35dd3bd1cec8d490f88173cf417e0400f9e771fabbcd0ce80fe3) for fetching information about batch of replica shards.
* Similar pull request for primary shard allocator transport action https://github.com/opensearch-project/OpenSearch/pull/8218

* Tested the changes manually by changing the replica shard allocator in the gateway allocator in my custom branch https://github.com/sudarshan-baliga/OpenSearch/commit/0919259abcb2b3aa9c85d0b69cbeca2818c36b1e and verified that the batch of shard information was fetched  while adding and removing the node from the cluster.

Created new index 
```
curl -X PUT "localhost:9200/my-index-4?pretty" -H 'Content-Type: application/json' -d'
{
  "settings": {
    "index": {
      "number_of_shards": 2,
      "number_of_replicas": 1
    }
  }
}
'
```
Removed the data2 node and re-added it.
Data node log which was dropped and re-added. We can see that batch of shard information was fetched instead of one shard.
```
 ] [data2] started
[2023-06-29T20:29:28,613][INFO ][o.o.i.s.TransportNodesBatchListShardStoreMetadata] [data2] Loaded store meta data for {[my-index-4][0]=[[StoreFilesMetadata{, shardId=[my-index-4][0], metadataSnapshot{size=1, syncId=null}}]], [my-index-4][1]=[[StoreFilesMetadata{, shardId=[my-index-4][1], metadataSnapshot{size=1, syncId=null}}]]} shards
```

Master node log when the data node was readded. We can see that batch of shard information was fetched instead of one shard.
```
[2023-06-29T20:29:29,074][INFO ][o.o.g.G.TestInternalReplicaShardAllocator] [master1] sdarbStore replica data {{data1}{PSqG-WkGSD2WiCi3dvXccw}{J84l7dayS2mx-8ocMXMn_Q}{127.0.0.1}{127.0.0.1:9301}{dir}{shard_indexing_pressure_enabled=true}=[[{data1}{PSqG-WkGSD2WiCi3dvXccw}{J84l7dayS2mx-8ocMXMn_Q}{127.0.0.1}{127.0.0.1:9301}{dir}{shard_indexing_pressure_enabled=true}][StoreFilesMetadata{, shardId=[my-index-4][0], metadataSnapshot{size=1, syncId=null}}]], {data2}{EMGkWoq8Q2mK3zVXFyYK0g}{fW3xmnx7S16uJCiI9RjM9A}{127.0.0.1}{127.0.0.1:9302}{dir}{shard_indexing_pressure_enabled=true}=[[{data2}{EMGkWoq8Q2mK3zVXFyYK0g}{fW3xmnx7S16uJCiI9RjM9A}{127.0.0.1}{127.0.0.1:9302}{dir}{shard_indexing_pressure_enabled=true}][StoreFilesMetadata{, shardId=[my-index-4][0], metadataSnapshot{size=1, syncId=null}}]]}
[2023-06-29T20:29:29,438][INFO ][o.o.c.r.a.AllocationService] [master1] Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[my-index-4][1], [my-index-4][0]]]).
[2023-06-29T20:29:29,734][INFO ][o.o.g.GatewayAllocator   ] [master1] sdarbStore Collecting of total shards ={}, over transport
```

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
* https://github.com/opensearch-project/OpenSearch/issues/8957

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/gateway/AsyncShardsFetchPerNode.java
Comment (Gaurav614) [inline server/src/main/java/org/opensearch/gateway/AsyncShardsFetchPerNode.java:27]: In your [PR](https://github.com/opensearch-project/OpenSearch/pull/8218) you are already created this class. This will have merge conflicts.

Please keep it in a single PR. 
Since we are going with Batch terminology, Let rename the class to `AsyncBatchShardFetch`

**Comment by Gaurav614** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> 
>     * **RESULT:** FAILURE ❌
> 
>     * **URL:** https://build.ci.opensearch.org/job/gradle-check/20265/
> 
>     * **CommitID:** [c02c4a0](https://github.com/opensearch-project/OpenSearch/commit/c02c4a0d5d102588116fe2def97f9f62af57dec1)
>       Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>       Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?


Please fix all the github actions failure. This PR cant be directly merged to main without those fixes

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (Gaurav614) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:429]: Earlier for StartedShards(PSA's transport action) we were storing per shard exception in the response. But for StoreShards(RSA's transport action) we were were directly throwing exception. With your changes now we will be storing exception per shard. Can you please verify what will be the implication of it in processing in master [here](https://github.com/opensearch-project/OpenSearch/blob/2.3/server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java#L200)

Scenario:
1. Just throw exception from data node over transport, and verify it is processed in List<T> responses rather than in failures condition. If this is true that you will not be changing any failure handling at master.

Comment (sudarshan-baliga) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:429]: [I induced the error on the data node](https://github.com/Gaurav614/OpenSearch/commit/78794d7ce5997bfe960499e4af659c81de7847ef#diff-ddb9033a1ecb36461c97bd2f4e06e8a18946c965f62ae4b3bdd15046edf8590a), it is processed in the onResponse as expected instead of onFailure case. It needs to be handled [like this](https://github.com/Gaurav614/OpenSearch/blob/async-shard-fetch-replica-induced-error/server/src/main/java/org/opensearch/gateway/GatewayAllocator.java#L526), I think this will be a separate task.

Log on Data 1 node
```
[2023-07-20T23:52:01,055][INFO ][o.o.i.s.TransportNodesListShardStoreMetadataBatch] [data1] sdarb Inducing error on node data1
```

Log on  Master node Gatewayallocator
```
[2023-07-20T23:53:08,826][ERROR][o.o.g.G.TestInternalReplicaShardAllocator] [master1] sdarb There was an error while fetching replica store data for shard [my-index-4][0], node[DQIxpsnNRA24tv_C4tkNOQ], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=H9uklZEHS3iUTPAnn6b86g], unassigned_info[[reason=NODE_LEFT], at[2023-07-20T18:20:23.323Z], delayed=false, details[node_left [Phd2Sj8gT26Kc3OvIxtm0w]], allocation_status[no_attempt]] on node {data1}{Phd2Sj8gT26Kc3OvIxtm0w}{l0V0RsyIRL62yNcYCbjudA}{127.0.0.1}{127.0.0.1:9302}{dir}{shard_indexing_pressure_enabled=true} so skipping this store data
```
Log on Data 1 node Master node AsyncShardFetch
```
[2023-07-20T23:53:07,958][INFO ][o.o.g.GatewayAllocator   ] [master1] Got the response
```

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (sudarshan-baliga) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:72]: Will fix it in next commit.

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (Gaurav614) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:429]: @sudarshan-baliga  Thanks. 
@shiv0408 Please make a note of while making changes for ReplicaShardAllocator

File: server/src/main/java/org/opensearch/gateway/AsyncShardsFetchPerNode.java
Comment (sudarshan-baliga) [reply server/src/main/java/org/opensearch/gateway/AsyncShardsFetchPerNode.java]: I have renamed and created the exact same file at both places to avoid merge conflicts and compilation failure in CI.

File: server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java
Comment (amkhar) [inline server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java:32]: Should we keep a single name everywhere ?

**shardsToCustomDataPath** ??

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (amkhar) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:147]: Do you think it'll be helpful to provide the details about output of this method ? may be convert this comment to standard method comment which includes request, response and exceptions.

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (amkhar) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:512]: just **storeFilesMetadataBatch** ?

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (Gaurav614) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:480]: nit: Documentation misleading here, how it used in TransportNodesAction. Or is it I am missing something here?

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (Gaurav614) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:509]: same as above

File: server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java
Comment (amkhar) [inline server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java:1]: Please keep in mind, that once https://github.com/opensearch-project/OpenSearch/pull/8218 is merged. You'll need to remove this class from this PR before merging.

**Review by amkhar — COMMENTED** [review]
Thanks @sudarshan-baliga

Please check if we can add an integ test to avoid code coverage failure.

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (Gaurav614) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:512]: I think what @sudarshan-baliga has kept is analogous to Primaries response, where we have prefix of `Node` and in replicas transport(without batch one) we have this prefix.

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (sudarshan-baliga) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:480]: As we are extending TransportNodesAction and the same class uses generics our specified classes for request and response will be used while performing  the transport call.
https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (sudarshan-baliga) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:509]: Replied in the previous comment.

File: server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java
Comment (sudarshan-baliga) [reply server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java:32]: Addressed in the new commit.

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (sudarshan-baliga) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:147]: Updated the documentation in the new commit.

File: server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java
Comment (sudarshan-baliga) [reply server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java:1]: File in both of the places are same, we won't face any merge conflict.

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (sudarshan-baliga) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:512]: I have kept the naming similar to the non batch version and same is the case with primary transpsort action .

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:61]: Same comments from other PR to not have code duplication - https://github.com/opensearch-project/OpenSearch/pull/8218#discussion_r1286694900

**Comment by kotwanikunal** [comment]
Apologies. This PR was auto closed without reaching a resolution from the maintainers. 
Re-opening to move it forward.
Thanks for your contributions to OpenSearch!

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (vikasvb90) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:62]: +1 on Shweta's comment. Are we planning this change?

File: server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java
Comment (amkhar) [inline server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java:1]: Remove this file. @shiv0408

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (amkhar) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:48]: Code duplication is removed by using helper class.

**Comment by amkhar** [comment]
Opened a new issue to improve the ShardId key size returned from the transport : https://github.com/opensearch-project/OpenSearch/issues/12010

File: server/src/main/java/org/opensearch/indices/store/StoreFilesMetadata.java
Comment (shiv0408) [reply server/src/main/java/org/opensearch/indices/store/StoreFilesMetadata.java]: Moved this class to Helper class

File: server/src/internalClusterTest/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatchIT.java
Comment (shiv0408) [reply server/src/internalClusterTest/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatchIT.java]: Moved the tests to RecoveryFromGatewayIT.java class as suggested

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (shiv0408) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java]: Catching these exceptions in line with how previously transport action was only catching the IOException and all the other kind of exceptions can cause the request to fail.

File: server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java
Comment (shiv0408) [reply server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java:1]: Removed

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (shiv0408) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:49]: Added a helper class to remove the duplication

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadata.java
Comment (amkhar) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadata.java:171]: @shiv0408 to maintain the old transport as it is, let's throw exception from here as it was being done. 
We only need modification(sending exception as part of response) in the new transport.

Comment (shiv0408) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadata.java:171]: We were not throwing exception before as well in case of IOException or IndexNotFoundException (which is anyway a subclass of IOExcpetion), so the behavior is not changed

https://github.com/opensearch-project/OpenSearch/blob/76ae14a4f2e99592610d3181543f5036214ceb7a/server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadata.java#L186-L191

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (amkhar) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:154]: @shiv0408  I think you would need to handle OpenSearchException also here. As you're running this loop, you would want it to complete for all the shards and now throw OpenSearchException in between.

Comment (shiv0408) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java]: Updated in the latest commit.

**Review by amkhar — APPROVED** [review]
Thanks @shiv0408 for making last set of changes.

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadata.java
Comment (shiv0408) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadata.java:171]: @shwetathareja IndexNotFoundException is an extension of IOException. Therefore the above code, where we are just catching IOException will be sufficient to catch and return empty response in case of IndexNotFoundException.

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (shiv0408) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:154]: Got your point. Will update this to catch all exceptions so that batch does not fail.

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataHelper.java
Comment (shiv0408) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataHelper.java:51]: Done

**Comment by shiv0408** [comment]
Failing tests:
- AdmissionControlMultiNodeIT.testAdmissionControlRejectionOnDisabled - identified flaky test #12620
- SegmentReplicationSnapshotIT.testSnapshotOnDocRep_RestoreOnDocRep - test got stuck and test suite timed out

**Comment by shiv0408** [comment]
`org.opensearch.repositories.azure.AzureBlobContainerRetriesTests.testWriteLargeBlob`
Create a new issue #12651 for better tracking of this flaky test, this was already identified as flaky in #6090.

`org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testIndexReopenClose`
#10987

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java:242]: why do we need the BaseShardResponse class?

File: server/src/main/java/org/opensearch/gateway/BaseShardResponse.java
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/gateway/BaseShardResponse.java:21]: @amkhar why do we need this class?

Comment (amkhar) [reply server/src/main/java/org/opensearch/gateway/BaseShardResponse.java]: Removing it, as it'll be pushed as part of   https://github.com/opensearch-project/OpenSearch/pull/12504/files#diff-f22c9cb55b850db5c383ce5ebfa022765e7089d220e80bd216b208cc0e7864fcR315-R318

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java
Comment (amkhar) [reply server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java]: It has common functionality of providing exception faced at shard level, as well as to check if the shard level response is empty or not. Keeping this encapsulated help avoiding if/else for primary and replica in the cache code.

Removing it for now as it's not needed in this change. It'll be pushed as part of [https://github.com/opensearch-project/OpenSearch/pull/12504](https://github.com/opensearch-project/OpenSearch/pull/12504)

**Comment by amkhar** [comment]
> ❌ Gradle check result for [2cc687f](https://github.com/opensearch-project/OpenSearch/commit/2cc687f2e420758f1da43775d75dc7e6edc3b3a6): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/34872/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test : testIsolateClusterManagerAndVerifyClusterStateConsensus https://github.com/opensearch-project/OpenSearch/issues/12095

**Review by shwetathareja — APPROVED** [review]

**Pull request merged**
